src/rdkafka_transport.c

```diff
@@ -901,6 +1115,27 @@ int rd_kafka_transport_ssl_ctx_init (rd_kafka_t *rk,
 		}
 	}
 
+    /* Register the peer verify callback on the context */
+    if (rk->rk_conf.ssl.ssl_cert_verify_cb) {
+
+        SSL_CTX_set_verify(
+            ctx, SSL_VERIFY_PEER | SSL_VERIFY_CLIENT_ONCE | SSL_VERIFY_FAIL_IF_NO_PEER_CERT,
+            NULL);
+
+        rd_kafka_dbg(rk, SECURITY, "SSL",
+            "Registering verification callback");
+
+        SSL_CTX_set_cert_verify_callback(
```

Comment:
Reading the SSL docs on [SSL_CTX_set_verify](https://www.openssl.org/docs/man1.0.2/man3/SSL_CTX_set_verify.html) and [SSL_CTX_set_cert_verify](https://www.openssl.org/docs/man1.0.2/man3/SSL_CTX_set_cert_verify_callback.html) makes me believe we should not do the latter, but the former.
The latter replaces OpenSSL's verification mechanism, which I don't think we want, while the former allows additional checks after the OpenSSL verification has been done.

Do you have a use-case for replicating the OpenSSL verification, or are the simpler additional verification sufficient?

Also see answers on SO: https://stackoverflow.com/questions/2729254/ssl-ctx-set-cert-verify-callback-vs-ssl-ctx-set-verify

**Do not push any commits to this PR**, I'm working on it locally with a bit of merge fixes, refactoring, etc.

Reply:
My understanding of this is that we do not replace OpenSSL's verification since we pass NULL to the callback in SSL_CTX_set_verify.  We are only specifying the verification flags.  Since we specify the flag SSL_VERIFY_PEER we need to use SSL_CTX_set_cert_verify_callback to validate the server to perform mutual auth.  

Looking at this PR we may want to add a property to set the verification flags so that somone can specify the flags they want.  In some cases SSL_VERIFY_FAIL_IF_NO_PEER_CERT may not be appropriate.

Reply:
My understanding of this is that we do not replace OpenSSL's verification since we pass NULL to the callback in SSL_CTX_set_verify.  We are only specifying the verification flags.  Since we specify the flag SSL_VERIFY_PEER we need to use SSL_CTX_set_cert_verify_callback to validate the server to perform mutual auth.  

Looking at this PR we may want to add a property to set the verification flags so that someone can specify the flags they want.  In some cases SSL_VERIFY_FAIL_IF_NO_PEER_CERT may not be appropriate.

Reply:
SSL_CTX_set_cert_verify() will replace OpenSSL's verification code according to [docs](https://www.openssl.org/docs/man1.0.2/man3/SSL_CTX_set_cert_verify_callback.html): 

> . If a verification callback callback is specified via SSL_CTX_set_cert_verify_callback(), the supplied callback function is called instead. 

I don't think we should do this, we should use SSL_CTX_set_verify() instead that allows an application to provide *additional* checks.

Also, we should have the verification enabled by default, regardless if a custom SSL verify callback is provided, i.e., we always have SSL_VERIFY_PEER set. 
The other flags are server-side only.

Reply:
I agree that SSL_CTX_set_cert_verify() will override the default cert checking mechanism.
I couldn't find confirmation, but is something like OCSP or another custom CA downloading and checking system may be ligitimately able to be used?

@edenhill Is your concern leaving it implemented technical or security in nature?

Reply:
@edenhill 

I agree that we definitely do not want to override the default certificate verification process.  We do however want to provide a way for the client to perform mutual auth based on the server/broker certificate.  If I remember correctly I could not get the cert in the SSL_CTX_set_verify callback.  When I called X509_STORE_CTX_get_current_cert this would only return the 0 dept certificate if there was an error in the validation process.  Otherwise NULL was returned.

I am ok with using this as long as there is a reliable way to retrieve the server certificate.

Reply:
Why conditionally perform mutual auth depending on the peer cert?
If you have an expected CA file/path configured, then openssl will already look for and verify the cert chain by default.

I believe that you'd use SSL_CTX_set_verify_callback to verify a peer cert against a custom store, e.g. a custom Windows CAPI store.

Reply:
@LighthouseJ

We need to verify a cert in the windows cert store.  In our case we have custom logic to verify the certificate in addition to it existing in the windows cert store.  Do you have any suggestions on how to handle this scenario?

Reply:
@noahdav If you need custom logic aside from the cert belonging in the default places for Windows, then you would likely need a custom verify callback function then.  IMHO I think your use case then necessitates the openssl callback function being exposed through librdkafka.

Reply:
@LighthouseJ That is why I created this PR in the first place.  This is being used internally in Microsoft Azure.

Reply:
@noahdav I'll see if we can get the full cert chain using the simpler callback form.

Reply:
@edenhill That would be great.  Let me know what you figure out.  I do think it would be best to not override the default verification but just plug the callback logic when we try to allow additional checks on top of what OpenSSL already does.

Reply:
@edenhill is there any update on this?  Were you able to get the entire cert chain in a different way?